libs/tdigest/src/main/java/org/elasticsearch/tdigest/arrays/TDigestDoubleArray.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ * This project is based on a modification of https://github.com/tdunning/t-digest which is licensed under the Apache 2.0 License.
+ */
+
+package org.elasticsearch.tdigest.arrays;
+
+/**
+ * Minimal interface for DoubleArray-like classes used within TDigest.
+ */
+public interface TDigestDoubleArray {
+    int size();
+
+    double get(int index);
+
+    void set(int index, double value);
+
+    void add(double value);
+
+    void ensureCapacity(int requiredCapacity);
+
+    /**
+     * Resizes the array. If the new size is bigger than the current size, the new elements are set to 0.
+     */
+    void resize(int newSize);
+
+    /**
+     * Copies {@code len} elements from {@code buf} to this array.
+     * <p>
+     *     Copy must be made in reverse order. That is, starting from offset+len-1 to offset.
+     *     This is, because it will be used to copy an array to itself.
```

Comment:
Nit: this statement is somewhat misleading, unless I'm mistaken. Are you referring to a specific case where self-copies are safe? If so, let's clearly document it and add asserts for it.

Reply:
I extracted those interfaces from existing TDigest usecases. Very specific, avoiding dead code.

This comes from a `System.arraycopy(src, srcIndex, dest, destIndex, len)`, which lets you have the same source and destination, and will ensure that it will be copied correctly in any case.
But as I didn't want to add dead code, I added that comment.

As to make it clearer, just pushed some improved comments (I believe), and some asserts so it's always called as it's expected to be called.
- For the Double array, just removed the comment and asserted that it can't be copied to itself
- For the Int array, asserted that if it's copied to itself, the source index should be greater

_Yeah, oddly specific..._